src/tdigest.rs

```diff
@@ -0,0 +1,269 @@
+// fn scale(q: f32, delta: f32) -> f32 {
+//     if q >= 0.5 {
+//         delta - delta * (2.0 - 2.0 * q).sqrt() / 2.0
+//     } else {
+//         delta * (2.0 * q).sqrt() / 2.0
+//     }
+// }
+
+// inverse of the scale function (commented above)
+fn inverse_scale(k: f32, delta: f32) -> f32 {
+    let kd_ratio = k / delta; // funny gamer joke. see the reason it's funny is because
+    if kd_ratio >= 0.5 {
+        1.0 - 2.0 * f32::powi(1.0 - kd_ratio, 2)
+    } else {
+        2.0 * f32::powi(kd_ratio, 2)
+    }
+}
+
+#[derive(Clone, Copy)]
+struct Centroid {
+    mean: f32,
+    weight: i32,
+}
+
+impl Centroid {
+    fn new(mean: f32, weight: i32) -> Self {
+        Self {
+            mean: mean,
+            weight: weight,
+        }
+    }
+
+    // combines two centroids. probably room for optimization here
+    fn merge(&mut self, c: &Centroid) {
+        let sum_weights = self.weight + c.weight;
+        let w1 = self.weight as f32 / sum_weights as f32;
+        let w2 = c.weight as f32 / sum_weights as f32;
+
+        let mean = w1 * self.mean + w2 * c.mean;
+        let weight = self.weight + c.weight;
+
+        self.mean = mean;
+        self.weight = weight;
+    }
+}
+
+/// T-Digest algorithm. A probabilistic data structure for estimating quantiles of data while using a fixed amount of memory.
+pub struct TDigest {
+    centroids: Vec<Centroid>,
+    delta: f32,
+    buffer: [f32; 32], // used exclusively for add_value
+    index: u8, // see above comment
+    temp_centroids: Vec<Centroid>, // used exclusively for cluster_centroids to save on vector allocations
+}
+
+impl TDigest {
+    /// Creates a new empty T-Digest given a delta value.
+    pub fn new_empty(delta: f32) -> Self {
+        Self {
+            centroids: Vec::new(),
+            delta,
+            buffer: [0.0; 32],
+            index: 0,
+            temp_centroids: Vec::new(),
+        }
+    }
+
+    // the following is not necessary for the python package (you can only have one new function)
+    /// Creates a new T-Digest from an input vector of unsorted values and a delta value. Equivalent to creating an empty T-Digest and merging in a vector of values.
+    // pub fn new_from_vec(v: Vec<f32>, delta: f32) -> Self {
+    //     let mut digest = Self {
+    //         centroids: Vec::new(),
+    //         delta,
+    //         buffer: [0.0; 32],
+    //         index: 0,
+    //         temp_centroids: Vec::new(),
+    //     };
+    //     digest.merge_vec_unsorted(v);
+    //     digest
+    // }
+
+    // apparently, this is not a std function
```

Comment:
still feels nicer to me to just write this function rather than drag in an entire dependency to only use one part of it
i'll get rid of the comment though

Reply:
This is perfectly fine, but if you don't like it, another way you could write it is `v.max(lo).min(hi)`. Also, looking at the source code of the implementation in `num` [here](https://docs.rs/num-traits/0.2.15/src/num_traits/lib.rs.html#408), it's probably good to add an assert to check that `lo <= hi`:
```rust
fn clamp(v: f32, lo: f32, hi: f32) -> f32 {
    debug_assert!(lo <= hi, "lo must be less than or equal to hi");
    ...
}
```

Reply:
added this suggestion. the function was only there because i copied it (along with `estimate_value`) from a different crate after realizing that if i wrote `estimate_value` myself it would end up being fundamentally the same code.